data-prepper-plugins/s3-source/src/main/java/com/amazon/dataprepper/plugins/source/codec/CSVCodec.java

```diff
@@ -0,0 +1,152 @@
+                    .withData(parsedLine)
+                    .build();
+            eventConsumer.accept(new Record<>(event));
+        } catch (final CsvReadException csvException) {
+            LOG.error("Invalid CSV row, skipping this line. Consider using the CSV Processor if there might be inconsistencies " +
+                    "in the number of columns because it is more flexible. Error ", csvException);
+        } catch (JsonParseException jsonException) {
+            LOG.error("A JsonParseException occurred while reading a row of the CSV file, skipping line. Error ", jsonException);
+        } catch (final Exception e) {
+            LOG.error("An Exception occurred while reading a row of the CSV file. Error ", e);
+        }
+    }
+
+    private int getSizeOfFirstLine(final String firstLine) {
+        final CsvMapper firstLineMapper = new CsvMapper();
+        firstLineMapper.enable(CsvParser.Feature.WRAP_AS_ARRAY); // allows firstLineMapper to read with empty schema
+        char delimiterAsChar = this.config.getDelimiter().charAt(0);
+        char quoteCharAsChar = this.config.getQuoteCharacter().charAt(0);
+
+        final CsvSchema getFirstLineLengthSchema = CsvSchema.emptySchema().withColumnSeparator(delimiterAsChar).
+                withQuoteChar(quoteCharAsChar);
+
+        try (final MappingIterator<List<String>> firstLineIterator = firstLineMapper.readerFor(List.class).with(getFirstLineLengthSchema)
+                .readValues(firstLine)) {
+            final List<String> parsedFirstLine = firstLineIterator.nextValue();
+
+            return parsedFirstLine.size();
+        } catch (final IOException e) {
+            LOG.error("An exception occurred while reading first line", e);
+            return 0;
+        }
+    }
+
+    private CsvSchema createCsvSchemaFromConfig(final int firstLineSize) {
+        final List<String> userSpecifiedHeader = config.getHeader();
+        final List<String> actualHeader = new ArrayList<>();
+        final char delimiter = config.getDelimiter().charAt(0);
+        final char quoteCharacter = config.getQuoteCharacter().charAt(0);
+        int providedHeaderColIdx = 0;
+        for (; providedHeaderColIdx < userSpecifiedHeader.size() && providedHeaderColIdx < firstLineSize; providedHeaderColIdx++) {
+            actualHeader.add(userSpecifiedHeader.get(providedHeaderColIdx));
+        }
+        for (int remainingColIdx = providedHeaderColIdx; remainingColIdx < firstLineSize; remainingColIdx++) {
+            actualHeader.add(generateColumnHeader(remainingColIdx));
+        }
+        CsvSchema.Builder headerBuilder = CsvSchema.builder();
+        for (String columnName : actualHeader) {
+            headerBuilder = headerBuilder.addColumn(columnName);
+        }
+        CsvSchema schema = headerBuilder.build().withColumnSeparator(delimiter).withQuoteChar(quoteCharacter);
+        return schema;
+    }
+
+    private String generateColumnHeader(final int colNumber) {
```

Comment:
There's a good amount of shared logic between this class and the [CSV Processor](https://github.com/opensearch-project/data-prepper/blob/main/data-prepper-plugins/csv-processor/src/main/java/com/amazon/dataprepper/plugins/processor/csv/CSVProcessor.java#L91). Could we move some of that shared logic into its own class? Is there a place that make sense to put that?

@asifsmohammed looks like you created an issue for a similar problem https://github.com/opensearch-project/data-prepper/issues/1643

Reply:
+1 and I also noticed this while developing tests — a lot of the code from the Newline and JSON codec unit tests was applicable to the CSV codec unit tests. Sadly I'm not sure where this common code should live. I remember that David Venable suggested packaging the CSV codec and processor together, but I'm not sure how this would look with the CSV processor being its own plugin and the CSV codec attached to S3.

Reply:
@finnroblin , I believe once we make a generic codec solution (#1532) then we can have one project which has both. As it is now, this code has to reside in the S3 plugin. I'm fine to refactor this later.